### PR TITLE
Gas optimizations

### PR DESCRIPTION
**Pull request by mariogutval**
Mandatory:

- [x] Remove initializations to default values
- [x] Uncheck index variable increment in for-loops
- [x] Use calldata instead of memory wherever is possible
- [x] Check require strings length

Optional:

- [x] Substitute require statements with custom errors

### CODE REVIEW

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 1]
```diff
@@ -425,14 +425,15 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
         }
         uint256 mostRecent = 0;
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Missing here.

**Review by diego-G — CHANGES_REQUESTED** [review]

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 2]
```diff
@@ -81,14 +93,13 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
     */
     function _scheduleSnapshot(uint256 time) internal {
         // Check the time firstly to avoid an useless read of storage
-        require(time > block.timestamp, "Snapshot scheduled in the past");
+        if(!(time > block.timestamp)) revert SnapshotScheduledInThePast(time, block.timestamp);
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Easier to read `if(time <= block.timestamp)`

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 3]
```diff
@@ -81,14 +93,13 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
...
 
         if (_scheduledSnapshots.length > 0) {
             // We check the last snapshot on the list
-            require(
-                time > _scheduledSnapshots[_scheduledSnapshots.length - 1],
-                "time has to be greater than the last snapshot time"
-            );
+            if(!(time > _scheduledSnapshots[_scheduledSnapshots.length - 1])) {
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same as above.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 4]
```diff
@@ -98,10 +109,10 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
     @dev schedule a snapshot at the specified time
     */
     function _scheduleSnapshotNotOptimized(uint256 time) internal {
-        require(time > block.timestamp, "Snapshot scheduled in the past");
+        if(!(time > block.timestamp)) revert SnapshotScheduledInThePast(time, block.timestamp);
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same as above.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 5]
```diff
@@ -125,25 +136,19 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
     */
     function _rescheduleSnapshot(uint256 oldTime, uint256 newTime) internal {
         // Check the time firstly to avoid an useless read of storage
-        require(oldTime > block.timestamp, "Snapshot already done");
-        require(newTime > block.timestamp, "Snapshot scheduled in the past");
-        require(_scheduledSnapshots.length > 0, "no scheduled snapshot");
+        if(!(oldTime > block.timestamp)) revert SnapshotAlreadyDone();
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 6]
```diff
@@ -125,25 +136,19 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
...
+        if(!(newTime > block.timestamp)) revert SnapshotScheduledInThePast(newTime, block.timestamp);
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 7]
```diff
@@ -125,25 +136,19 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
...
-            );
+            if(!(newTime < _scheduledSnapshots[index + 1])) revert SnapshotTimestampAfterNextSnapshot(newTime, _scheduledSnapshots[index + 1]);
         }
 
         if (index > 0) {
-            require(
-                newTime > _scheduledSnapshots[index - 1],
-                "time has to be greater than the previous snapshot"
-            );
+            if(!(newTime > _scheduledSnapshots[index - 1])) revert SnapshotTimestampBeforePreviousSnapshot(newTime, _scheduledSnapshots[index - 1]);
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 8]
```diff
@@ -156,13 +161,10 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
     */
     function _unscheduleLastSnapshot(uint256 time) internal {
         // Check the time firstly to avoid an useless read of storage
-        require(time > block.timestamp, "Snapshot already done");
-        require(_scheduledSnapshots.length > 0, "No snapshot scheduled");
+        if(!(time > block.timestamp)) revert SnapshotAlreadyDone();
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same.

File: contracts/modules/internal/SnapshotModuleInternal.sol [hunk 9]
```diff
@@ -174,9 +176,9 @@ abstract contract SnapshotModuleInternal is ERC20Upgradeable {
     - Reduce the array size by deleting the last snapshot
     */
     function _unscheduleSnapshotNotOptimized(uint256 time) internal {
-        require(time > block.timestamp, "Snapshot already done");
+        if(!(time > block.timestamp)) revert SnapshotAlreadyDone();
```
Comment (diego-G) [inline contracts/modules/internal/SnapshotModuleInternal.sol]: Same.

File: contracts/modules/wrapper/optional/ValidationModule.sol [hunk 10]
```diff
@@ -59,7 +61,7 @@ abstract contract ValidationModule is
     function setRuleEngine(
         IEIP1404Wrapper ruleEngine_
     ) external onlyRole(DEFAULT_ADMIN_ROLE) {
-        require(ruleEngine != ruleEngine_, "Same value");
+        if(ruleEngine == ruleEngine_) revert SameValue();
```
Comment (diego-G) [inline contracts/modules/wrapper/optional/ValidationModule.sol]: Why not `_ruleEngine` ?

File: contracts/modules/security/AuthorizationModule.sol [hunk 11]
```diff
@@ -43,7 +45,7 @@ abstract contract AuthorizationModule is AccessControlUpgradeable {
     function __AuthorizationModule_init_unchained(
         address admin
     ) internal onlyInitializing {
-        require(admin != address(0), "Address 0 not allowed");
+        if(admin == address(0)) revert InvalidAddress();
```
Comment (diego-G) [inline contracts/modules/security/AuthorizationModule.sol]: Rename the error to `AddressZeroNotAllowed`.

File: contracts/libraries/Errors.sol [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+pragma solidity ^0.8.17;
+    
+error BeforeTokenTransfer(address from, address to, uint256 amount);
+error SnapshotScheduledInThePast(uint256 time, uint256 timestamp);
+error SnapshotTimestampBeforeLastSnapshot(uint256 time, uint256 lastSnapshotTimestamp);
+error SnapshotTimestampAfterNextSnapshot(uint256 time, uint256 nextSnapshotTimestamp);
+error SnapshotTimestampBeforePreviousSnapshot(uint256 time, uint256 previousSnapshotTimestamp);
+error SnapshotAlreadyExists();
+error SnapshotAlreadyDone();
+error NoScheduledSnapshot();
```
Comment (diego-G) [inline contracts/libraries/Errors.sol]: Rename to `SnapshotNotScheduled` to follow pattern.

File: contracts/libraries/Errors.sol [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+error SnapshotNotFound();
+error NotLastSnapshot();
```
Comment (diego-G) [inline contracts/libraries/Errors.sol]: Rename to `SnapshotNeverScheduled`.

File: contracts/libraries/Errors.sol [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+error InvalidAddress();
+error DirectCallToImplementation();
+error Allowance(uint256 allowance, uint256 currentAllowance);
+error SameValue();
+error SetRuleEngine(address ruleEngine);
```
Comment (diego-G) [inline contracts/libraries/Errors.sol]: This error is never used. Please, remove it.

File: contracts/libraries/Errors.sol [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+error SnapshotTimestampAfterNextSnapshot(uint256 time, uint256 nextSnapshotTimestamp);
+error SnapshotTimestampBeforePreviousSnapshot(uint256 time, uint256 previousSnapshotTimestamp);
+error SnapshotAlreadyExists();
+error SnapshotAlreadyDone();
+error NoScheduledSnapshot();
+error SnapshotNotFound();
+error NotLastSnapshot();
+error InvalidAddress();
+error DirectCallToImplementation();
+error Allowance(uint256 allowance, uint256 currentAllowance);
```
Comment (diego-G) [inline contracts/libraries/Errors.sol]: Rename to `WrongAllowance`.

File: contracts/libraries/Errors.sol [hunk 16]
```diff
@@ -0,0 +1,19 @@
+//SPDX-License-Identifier: MPL-2.0
+
+pragma solidity ^0.8.17;
+    
+error BeforeTokenTransfer(address from, address to, uint256 amount);
```
Comment (diego-G) [inline contracts/libraries/Errors.sol]: Rename to `InvalidTransfer`.

**Review by diego-G — CHANGES_REQUESTED** [review]
Minimal changes.

File: contracts/modules/wrapper/optional/ValidationModule.sol (hunk 10 above)
Comment (mariogutval) [reply contracts/modules/wrapper/optional/ValidationModule.sol]: I just let it as it was

File: contracts/modules/internal/SnapshotModuleInternal.sol (hunk 2 above)
Comment (mariogutval) [reply contracts/modules/internal/SnapshotModuleInternal.sol]: There is no gas optimization, so I'll use your way, which is more readable

**Review by diego-G — APPROVED** [review]

**Pull request merged**